core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -3956,11 +3961,482 @@ class KafkaApis(val requestChannel: RequestChannel,
+      case Some(manager) => manager
+      case None => throw new IllegalStateException("ShareFetchRequest received but SharePartitionManager is not initialized")
+    }
+
+    val groupId = shareFetchRequest.data.groupId
+    val clientId = request.header.clientId
+    val memberId = shareFetchRequest.data().memberId()
+    val shareSessionEpoch = shareFetchRequest.data().shareSessionEpoch()
+
+    val shareFetchData = shareFetchRequest.shareFetchData(topicNames)
+    val forgottenTopics = shareFetchRequest.forgottenTopics(topicNames)
+    var cachedTopicPartitions : util.List[TopicIdPartition] = null
+
+    if (shareSessionEpoch == ShareFetchMetadata.FINAL_EPOCH) {
+      try {
+        cachedTopicPartitions = sharePartitionManager.cachedTopicIdPartitionsInShareSession(groupId, Uuid.fromString(memberId))
+      } catch {
+        // Exception handling is needed when this value is being utilized on receiving FINAL_EPOCH.
+        case _: ShareSessionNotFoundException => cachedTopicPartitions = null
+      }
+    }
+
+    def isAcknowledgeDataPresentInFetchRequest() : Boolean = {
+      var isAcknowledgeDataPresent = false
+      shareFetchRequest.data().topics().forEach ( topic => {
+        breakable{
+          topic.partitions().forEach ( partition => {
+            if (partition.acknowledgementBatches() != null && !partition.acknowledgementBatches().isEmpty) {
+              isAcknowledgeDataPresent = true
+              break()
+            } else {
+              isAcknowledgeDataPresent = false
+            }
+          })
+        }
+      })
+      isAcknowledgeDataPresent
+    }
+
+    val isAcknowledgeDataPresent = isAcknowledgeDataPresentInFetchRequest()
+    var shareFetchResponse : ShareFetchResponse = null
+    // Variable to store any error thrown while the handling piggybacked acknowledgements
+    var acknowledgeError : Errors = Errors.NONE
+    // Variable to store the topic partition wise result of piggybacked acknowledgements
+    var acknowledgeResult = mutable.Map[TopicIdPartition, ShareAcknowledgeResponseData.PartitionData]()
+
+    def isInvalidShareFetchRequest() : Boolean = {
+      // The Initial Share Fetch Request should not Acknowledge any data
+      if (shareSessionEpoch == ShareFetchMetadata.INITIAL_EPOCH && isAcknowledgeDataPresent) {
+        return true
+      }
+      false
+    }
+
+    val newReqMetadata : ShareFetchMetadata = new ShareFetchMetadata(Uuid.fromString(memberId), shareSessionEpoch)
+    var shareFetchContext : ShareFetchContext = null
+    try {
+      // Creating the shareFetchContext for Share Session Handling
+      shareFetchContext = sharePartitionManager.newContext(groupId, shareFetchData, forgottenTopics, newReqMetadata)
+    } catch {
+      case e: Exception => shareFetchResponse = shareFetchRequest.getErrorResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, e) match {
+        case response: ShareFetchResponse => response
+        case _ => null
```

Comment:
I am not sure what does this case means i.e. is this check required? If yes, then what handling is done when `ShareFetchResponse` is not returned by the method call i.e. how it further gets handled?

Shouldn't we complete the request and return right away?

Reply:
Hi, thanks for the review. Yes this check is required here, because `sharePartitionManager.newContext` might throw errors during its execution. Initially, `shareFetchResponse` is defined as a null there. If an error is thrown, then its value is set as an error response. After this check, the acknowledging and fetching only proceeds if `shareFetchResponse` is null. If it is not null, then a final `ShareFetchResponse` object is prepared with the error code and appropriate values for the acknowledgements and sent back to the user

Reply:
My concern is on below code, should we complete the API call if `errorResponse` is constructed?

```
case e: Exception => shareFetchResponse = shareFetchRequest.getErrorResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, e) match {
        case response: ShareFetchResponse => response
        case _ => null

Reply:
I agree with @apoorvmittal10 , seeing the newContext function, the possible errors are `INVALID_REQUEST`, `SHARE_SESSION_NOT_FOUND` and `INVALID_SHARE_SESSION_EPOCH`. In all such cases, we should be completing the API call with a top level there itself.

Reply:
That makes sense. Thanks for the review, will make the changes in the next commit